extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,202 @@
+
+  /**
+   * @param {!./amp-a4a.AmpA4A} a4a The AmpA4A instance to be refreshed.
+   * @param {!RefreshConfig} config
+   */
+  constructor(a4a, config) {
+
+    /** @const @private {!./amp-a4a.AmpA4A} */
+    this.a4a_ = a4a;
+
+    /** @const @private {!Window} */
+    this.win_ = a4a.win;
+
+    /** @const @private {!Element} */
+    this.element_ = a4a.element;
+
+    /** @const @private {string} */
+    this.adType_ = this.element_.getAttribute('type');
+
+    /** @const @private {?number} */
+    this.refreshInterval_ = this.getPublisherSpecifiedRefreshInterval_();
+
+    /** @const @private {!RefreshConfig} */
+    this.config_ = this.getConfiguration_(config);
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = timerFor(this.win_);
+
+    /** @private {?(number|string)} */
+    this.refreshTimeoutId_ = null;
+
+    /** @private {boolean} */
+    this.isRefreshable_ = !!(this.config_  // The network has opted-in.
+        // The publisher has enabled refresh on this slot.
+        && (this.refreshInterval_ || this.refreshInterval_ == '')
+        // The slot is contained only within container types eligible for
+        // refresh.
+        && !getEnclosingContainerTypes(this.element_).filter(container =>
+          container != ValidAdContainerTypes['AMP-CAROUSEL']
+          && container != ValidAdContainerTypes['AMP-STICKY-AD']).length);
+  }
+
+
+  /**
+   * Initiates the refresh cycle by initiating the visibility manager on the
+   * element.
+   *
+   * @return {?Promise} Promise that resolves after refresh event. This promise
+   *   is useful to have for test purposes.
+   */
+  initiateRefreshCycle() {
+    if (!this.isRefreshable()) {
+      return null;
+    }
+    return new Promise(resolve => {
+      analyticsForDoc(this.element_, true).then(analytics => {
+        analytics.getAnalyticsRoot(this.element_).getVisibilityManager()
+            .listenElement(this.element_, this.config_, null, null, () => {
+              this.refreshTimeoutId_ = this.timer_.delay(() => {
+                this.a4a_.refresh(() => this.initiateRefreshCycle());
+                resolve();
+              }, /** @type {number} */ (this.refreshInterval_));
+            });
+      });
+    });
+  }
+
+  /**
+   * Terminates the current refresh cycle, if one currently exists. Returns
+   * true if the cycle was canceled successfully, false otherwise.
+   */
+  stopRefreshCycle() {
+    if (this.refreshTimeoutId_) {
+      this.timer_.cancel(this.refreshTimeoutId_);
+      this.refreshTimeoutId_ = null;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Converts config to appropriate units.
+   * @return {!RefreshConfig}
+   */
+  getConfiguration_(config) {
+    // Convert seconds to milliseconds.
+    config['totalTimeMin'] *= 1000;
+    config['continuousTimeMin'] *= 1000;
+    return config;
+  }
+
+  /**
+   * Retrieves the publisher-specified refresh interval, if one were set.
```

Comment:
Have comment reflect that we look first for attribute and then page level via metatag